File: mayonez-base/src/main/java/slavsquatsuperstar/mayonez/graphics/renderer/GLRenderer.java
package slavsquatsuperstar.mayonez.graphics.renderer;

import slavsquatsuperstar.mayonez.Preferences;
import slavsquatsuperstar.mayonez.io.Assets;
import slavsquatsuperstar.mayonez.io.GLTexture;
import slavsquatsuperstar.mayonez.io.Shader;

import java.util.ArrayList;
import java.util.List;

public abstract class GLRenderer {

    // GPU Resources
    protected final List<RenderBatch> batches;
    protected final Shader shader;
    protected final int[] textureSlots; // support multiple texture IDs in batch

    public GLRenderer(String shaderFile) {
        this.batches = new ArrayList<>();
        this.shader = Assets.getShader(shaderFile);
        textureSlots = new int[Preferences.getMaxTextureSlots()];
        for (int i = 0; i < this.textureSlots.length; i++) this.textureSlots[i] = i; // ints 0-7
    }

    protected abstract void rebuffer();

    protected final RenderBatch getAvailableBatch(GLTexture texture, int zIndex) {
        for (RenderBatch batch : batches) {
            if (batch.hasRoom() && batch.getZIndex() == zIndex) { // has room for sprite (vertices)
                if (batch.hasTextureRoom() || batch.hasTexture(texture) || texture == null) // has room for texture or using color
                    return batch;
            }
        }
        // All batches full
        RenderBatch batch = createBatch(zIndex);
        batch.clear();
        batches.add(batch);
        return batch;
    }

    protected abstract RenderBatch createBatch(int zIndex);

}

File: mayonez-base/src/main/kotlin/mayonez/renderer/gl/GLRenderer.kt
package mayonez.renderer.gl

import mayonez.*
import mayonez.graphics.*
import mayonez.renderer.*
import mayonez.renderer.batch.*
import mayonez.renderer.shader.*
import java.awt.*

/**
 * A base renderer for OpenGL that uploads sprite and shape data to the
 * GPU.
 *
 * @author SlavSquatSuperstar
 */
@UsesEngine(EngineType.GL)
abstract class GLRenderer(protected val shader: Shader) : Renderer {

    companion object {
        @JvmStatic
        protected val MAX_TEXTURE_SLOTS: Int = 8
    }

    // GPU Resources
    private val batches: MutableList<RenderBatch> = ArrayList()
    protected val textureSlots: IntArray = IntArray(MAX_TEXTURE_SLOTS) { it }

    // Renderer Methods

    override fun clear() {
        batches.forEach(RenderBatch::deleteBatch)
        batches.clear()
    }

    override fun render(g2: Graphics2D?) {
        // Re-buffer objects
        preRender() // Prepare batches and update uniforms
        batches.forEach(RenderBatch::clearVertices)
        createBatches()

        // Draw objects
        batches.sortBy(RenderBatch::getDrawOrder) // Sort batches by z-index
        batches.forEach {
            it.uploadUniforms(viewport, textureSlots)
            it.drawBatch()
        }
        postRender()
    }

    /** Clear the screen and upload resources to the GPU. */
    protected open fun preRender() {}

    /** Sort image data into render batches. */
    protected abstract fun createBatches()

    /** Finish drawing and free resources from the GPU. */
    protected open fun postRender() {
        shader.unbind() // Unbind everything
    }

    // Camera Methods

    // TODO maybe make field
    override fun getViewport(): Viewport = SceneManager.currentScene.camera

    // Batch Helper Methods

    /**
     * Uploads any necessary uniforms for the batch and draws its vertices.
     */
    abstract fun RenderBatch.uploadUniforms(viewport: Viewport, textureSlots: IntArray)

    // TODO see Cherno renderer class
    /**
     * Gets an available render batch for this object or creates a new one it all
     * batches are full.
     */
    protected fun GLRenderable.getAvailableBatch(): RenderBatch {
        val batch = batches.find { it.canFitObject(this) }
        if (batch != null) {
            return batch
        } else {
            val newBatch = this.createNewBatch()
            batches.add(newBatch)
            return newBatch
        }
    }

    /**
     * Creates a new render batch that can draw this object.
     */
    abstract fun GLRenderable.createNewBatch(): RenderBatch

}